WordPress/wordpress-playground
# php-wasm/node: Ship as ESM and CJS

The `@php-wasm/node` package used to be CommonJS only. This created tricky dependency identity problems in ESM applications, e.g. in #1577 the main script imports the ESM version of `@php-wasm/universal`, but `@php-wasm/node` imports the CJS version of it.

This PR ensures `@php-wasm/node` has an ESM version available.

This is first step towards solving https://github.com/WordPress/wordpress-playground/issues/1583

Closes https://github.com/WordPress/wordpress-playground/issues/1579 Closes https://github.com/WordPress/wordpress-playground/issues/1577

 ## Testing instructions

Run `npm run build`, copy the built packages from `dist/packages/php-wasm` to node_modules in another directory, create a test.js script with the following contents:

```
import { PHP } from '@php-wasm/universal';
import { loadNodeRuntime, useHostFilesystem } from '@php-wasm/node';

const php = new PHP(await loadNodeRuntime('8.0'));
console.log("Hey");
console.log(await php.run({ code: 'echo "Hello, World!";' }));
```

Then create a package.json file with `{"type":"module"}` and confirm that running test.js works. Then create test.cjs file as follows:

```js
const { PHP } = require('@php-wasm/universal');
const { loadNodeRuntime } = require('@php-wasm/node');

async function main() {
    const runtimeId = await loadNodeRuntime('8.0');
    console.log({runtimeId})
    const php = new PHP(runtimeId);

    console.log(php);
}
main();
```

And confirm this one works, too.
